### PR TITLE
Questions tests

### PR DESCRIPTION
**Pull request by pavkahanov**

### CODE REVIEW

File: contracts/ZeroOne/Questions/lib/Group.sol [hunk 1]
```diff
@@ -23,6 +23,7 @@ library GroupType {
         internal
         returns (bool changed)
     {
-        // TODO: implementation goes here
+         _self.name = _name;
```
Comment (masteralx) [inline contracts/ZeroOne/Questions/lib/Group.sol]: forgot changed result

File: contracts/ZeroOne/Questions/lib/GroupList.sol [hunk 2]
```diff
@@ -46,10 +46,9 @@ library GroupList {
         uint _id
     )
         internal
-        pure
+        view
         returns (bool valid)
     {
-        // TODO: implement this
-        return true;
+        return (_self.list.length - 1) >= _id;
```
Comment (masteralx) [inline contracts/ZeroOne/Questions/lib/GroupList.sol]: _id < _self.list.length

File: contracts/ZeroOne/Questions/lib/QuestionList.sol [hunk 3]
```diff
@@ -50,10 +50,9 @@ library QuestionList {
         uint _id
     )
         internal
-        pure
+        view
         returns (bool valid)
     {
-        // TODO: implement this
-        return true;
+        return (_self.list.length - 1) >= _id;
```
Comment (masteralx) [inline contracts/ZeroOne/Questions/lib/QuestionList.sol]: _id < _self.list.length

File: test/3_Questions.spec.js [hunk 4]
```diff
@@ -1,4 +1,5 @@
 const Questions = artifacts.require('./Questions.sol');
```
Comment (masteralx) [inline test/3_Questions.spec.js]: artifacts
-- line break --
usual requires

File: test/3_Questions.spec.js [hunk 5]
```diff
@@ -22,10 +23,16 @@ contract('Questions', (accounts) => {
 
     describe('constructor()', () => {
         it('should be successfully created', async () => {
-            // TODO: implement creation test
+            try {
```
Comment (masteralx) [inline test/3_Questions.spec.js]: remove try catch

Comment (masteralx) [reply test/3_Questions.spec.js]: add asserts for initial values

File: test/3_Questions.spec.js [hunk 6]
```diff
@@ -46,13 +53,91 @@ contract('Questions', (accounts) => {
                 }
             });
         });
+
+        it('shoud fail on uploading two same questions', async () => {
+            try {
```
Comment (masteralx) [inline test/3_Questions.spec.js]: add let error = false
set error = true on catch
assert error === true after catch

File: test/3_Questions.spec.js [hunk 7]
```diff
@@ -46,13 +53,91 @@ contract('Questions', (accounts) => {
...
+                questionWithIncorrectAddress.target = '';
+                await questions.addQuestion(questionWithIncorrectAddress);
+            } catch (e) {
+                assert.strictEqual(e.message, 'invalid address (arg="target", coderType="address", value="")')
+            }
+
+        });
+
+        it('should fail on uploading question with incorrect time limit', async () => {
+            try {
```
Comment (masteralx) [inline test/3_Questions.spec.js]: test both bounds

File: test/3_Questions.spec.js [hunk 8]
```diff
@@ -46,13 +53,91 @@ contract('Questions', (accounts) => {
...
+                questionWithIncorrectTime.timeLimit = 1;
+                await questions.addQuestion(questionWithIncorrectTime);
+            } catch (e) {
+                assert.strictEqual(e.message, getErrorMessage('Invalid question'));
+            }
+        });
+
+        it('should fail on uploading question with incorrect params', async () => {
+            try {
+                const questionWithIncorrectParams = Object.create(question);
```
Comment (masteralx) [inline test/3_Questions.spec.js]: test params length mismatch

File: test/3_Questions.spec.js [hunk 9]
```diff
@@ -46,13 +53,91 @@ contract('Questions', (accounts) => {
...
+        it('should fail on changing status on incorrect value ', async () => {
+            await questions.addQuestion(question);
+            await questions.setActiveStatus(0, 'test');
+            const uploaded = await questions.getQuestion(0);
+            assert.strictEqual(uploaded.active, true);
+        });
+    });
+
+
+    // : write tests for other cases
```
Comment (masteralx) [inline test/3_Questions.spec.js]: remove

File: test/4_QuestionsWithGroups.spec.js [hunk 10]
```diff
@@ -33,8 +34,90 @@ contract('QuestionsWithGroups', (accounts) => {
...
     });
 
+    describe('addQuestion()', () => {
+        it('should add question', async () => {
+            await questionsWithGroups.addQuestion(question);
+            const uploaded = await questionsWithGroups.getQuestion(0);
+            assert.strictEqual(uploaded.name, 'question name');
+        });
+
+        it('shoud fail on uploading two same questions', async () => {
```
Comment (masteralx) [inline test/4_QuestionsWithGroups.spec.js]: no need

File: test/4_QuestionsWithGroups.spec.js [hunk 11]
```diff
@@ -33,8 +34,90 @@ contract('QuestionsWithGroups', (accounts) => {
...
+        });
+    });
+
+    
+
+
     // TODO: write tests for other cases
-    // 1. test non-unique names
+    // * 1. test non-unique names
     // 2. test upload question and group with same names
```
Comment (masteralx) [inline test/4_QuestionsWithGroups.spec.js]: no test found

File: test/4_QuestionsWithGroups.spec.js [hunk 12]
```diff
@@ -33,8 +34,90 @@ contract('QuestionsWithGroups', (accounts) => {
...
+            } 
+        });
+    });
+
+    describe('getQuestionGroupsAmount()', () => {
+        it('should successful check amount of groups', async () => {
+            const group  = { name: 'something' };
+            await questionsWithGroups.addQuestionGroup(group);
+            const countOfUploaded = await questionsWithGroups.getQuestionGroupsAmount();
+            assert.equal(countOfUploaded, web3.utils.toBN(2));
```
Comment (masteralx) [inline test/4_QuestionsWithGroups.spec.js]: BN.toNumber()

File: test/4_QuestionsWithGroups.spec.js [hunk 13]
```diff
@@ -33,8 +34,90 @@ contract('QuestionsWithGroups', (accounts) => {
...
+            try {
+                await questionsWithGroups.getQuestionGroup(25);
+            } catch (e) {
+                assert.equal(e.message, 'Returned error: VM Exception while processing transaction: revert Provided index is out of bounds');
+            }
+        });
+    });
+
+    describe('addQuestionGroup()', () => {
+        it('should successful add group', async () => {
```
Comment (masteralx) [inline test/4_QuestionsWithGroups.spec.js]: successfully

File: test/4_QuestionsWithGroups.spec.js [hunk 14]
```diff
@@ -33,8 +34,90 @@ contract('QuestionsWithGroups', (accounts) => {
...
+    describe('setQuestionGroupName()', () => {
+        it('should successful change group name', async () => {
+            name = 'test'
+            await questionsWithGroups.addQuestionGroup(group);
+            await questionsWithGroups.setQuestionGroupName(1, name);
+            const uploadedGroup = await questionsWithGroups.getQuestionGroup(1);
+            assert.strictEqual(uploadedGroup.name, name);
+        });
+
+        it('should fail on change name of non-unique group', async ()=> {
```
Comment (masteralx) [inline test/4_QuestionsWithGroups.spec.js]: rename

File: test/3_Questions.spec.js [hunk 15]
```diff
@@ -46,13 +53,91 @@ contract('Questions', (accounts) => {
...
+            assert.strictEqual(uploaded.active, false);
+        });
+
+        it('should fail on changing status on incorrect value ', async () => {
+            await questions.addQuestion(question);
+            await questions.setActiveStatus(0, 'test');
+            const uploaded = await questions.getQuestion(0);
+            assert.strictEqual(uploaded.active, true);
+        });
+    });
```
Comment (masteralx) [inline test/3_Questions.spec.js]: add tests for events

File: test/4_QuestionsWithGroups.spec.js [hunk 16]
```diff
@@ -33,8 +34,90 @@ contract('QuestionsWithGroups', (accounts) => {
...
+            try {
+                name = 'test'
+                await questionsWithGroups.setQuestionGroupName(12, name);
+            } catch(e) {
+                assert.strictEqual(e.message, getErrorMessage('Provided index is out of bounds'));
+            }
+        });
+    });
```
Comment (masteralx) [inline test/4_QuestionsWithGroups.spec.js]: add tests for events

File: contracts/ZeroOne/Questions/lib/Group.sol (hunk 1 above)
Comment (pavkahanov) [reply contracts/ZeroOne/Questions/lib/Group.sol]: fixed

File: contracts/ZeroOne/Questions/lib/GroupList.sol (hunk 2 above)
Comment (pavkahanov) [reply contracts/ZeroOne/Questions/lib/GroupList.sol]: fixed

File: contracts/ZeroOne/Questions/lib/QuestionList.sol (hunk 3 above)
Comment (pavkahanov) [reply contracts/ZeroOne/Questions/lib/QuestionList.sol]: fixed

File: test/3_Questions.spec.js (hunk 4 above)
Comment (pavkahanov) [reply test/3_Questions.spec.js]: fixed

File: test/3_Questions.spec.js (hunk 6 above)
Comment (pavkahanov) [reply test/3_Questions.spec.js]: fixed

File: test/4_QuestionsWithGroups.spec.js (hunk 14 above)
Comment (pavkahanov) [reply test/4_QuestionsWithGroups.spec.js]: renamed

File: test/4_QuestionsWithGroups.spec.js (hunk 11 above)
Comment (pavkahanov) [reply test/4_QuestionsWithGroups.spec.js]: added

File: test/4_QuestionsWithGroups.spec.js (hunk 13 above)
Comment (pavkahanov) [reply test/4_QuestionsWithGroups.spec.js]: fixed

File: test/3_Questions.spec.js (hunk 15 above)
Comment (pavkahanov) [reply test/3_Questions.spec.js]: added

File: test/4_QuestionsWithGroups.spec.js (hunk 16 above)
Comment (pavkahanov) [reply test/4_QuestionsWithGroups.spec.js]: added

File: contracts/ZeroOne/Questions/QuestionsWithGroups.sol [hunk 17]
```diff
@@ -103,7 +103,7 @@ contract QuestionsWithGroups is Questions {
 
     /**
      * @notice adds new question to list
-     * @param _question question
+     * @param _question  question
```
Comment (masteralx) [inline contracts/ZeroOne/Questions/QuestionsWithGroups.sol]: extra space

File: contracts/ZeroOne/Questions/lib/Group.sol (hunk 1 above)
Comment (masteralx) [reply contracts/ZeroOne/Questions/lib/Group.sol]: return keccak256(abi.encodePacked(_self.name)) == keccak256(abi.encodePacked(_name))
            ? false
            : _self.name= _name;

File: contracts/__mocks__/SelectorMock.sol [hunk 18]
```diff
@@ -26,4 +26,15 @@ contract SelectorMock {
     {
         data = _selector.addData(_data);
     }
+
+    function testMakeCall(
```
Comment (masteralx) [inline contracts/__mocks__/SelectorMock.sol]: add docs

File: contracts/__mocks__/UniqueMock.sol [hunk 19]
```diff
@@ -0,0 +1,25 @@
+pragma solidity 0.6.1;
+
+import "../lib/UniqueNames.sol";
+
+
+/**
+ * @title SelectorMock
```
Comment (masteralx) [inline contracts/__mocks__/UniqueMock.sol]: mismatch

File: test/2_UniqueNames.spec.js [hunk 20]
```diff
@@ -1 +1,35 @@
...
+contract('Unique', (accounts) => {
+  let uniqueMock;
+  const deployFrom = accounts[0];
+
+  beforeEach(async () => {
+    uniqueMock = await UniqueMock.new({ from: deployFrom });
+  });
+
+  describe('add()', () => {
+    it('should add names to list', async () => {
```
Comment (masteralx) [inline test/2_UniqueNames.spec.js]: test isUnique method

File: contracts/__mocks__/UniqueMock.sol (hunk 19 above)
Comment (masteralx) [reply contracts/__mocks__/UniqueMock.sol]: method to check isUnuque

File: test/0_Meta.spec.js [hunk 21]
```diff
@@ -24,7 +28,21 @@ contract('Meta', (accounts) => {
...
+            const initial = (new Array(1)).fill(0);
+            try {
+                const value = randomInt(0, 100);
+                const replace = randomRange(0, 100, 2);
+                const base = web3.eth.abi.encodeParameters(failingParams, [...initial, value]);
+                const expected = web3.eth.abi.encodeParameters(params, [...replace, value]);
+                const result = await metaMock.testMeta(base, ...replace);
+                assert.strictEqual(result, expected);
+            } catch (e) {
+                error = true;
```
Comment (masteralx) [inline test/0_Meta.spec.js]: check for error message

File: test/1_Selector.spec.js [hunk 22]
```diff
@@ -30,8 +62,25 @@ contract('Selector', (accounts) => {
...
+    describe('makeCall()', () => {
+        it('should successfully call method from contract', async () => {
+            for (let i = 1; i < 10; i++) {
+                const value = `test ${i}`
+                const {address} = questionsWithGroups;
+                const encodedValue = web3.eth.abi.encodeParameters(['tuple(string)'], [[value]]);
+                await selectorMock.testMakeCall(addGroupSelector, address, encodedValue);
+                const group = await questionsWithGroups.getQuestionGroup(i);
+                assert.strictEqual(value, group.name);
+            }
```
Comment (masteralx) [inline test/1_Selector.spec.js]: test non-exisiting method call

File: test/3_Questions.spec.js (hunk 15 above)
Comment (masteralx) [reply test/3_Questions.spec.js]: re-check

File: test/4_QuestionsWithGroups.spec.js [hunk 23]
```diff
@@ -33,10 +35,87 @@ contract('QuestionsWithGroups', (accounts) => {
         });
     });
 
+    describe('getQuestionGroupsAmount()', () => {
+        it('should successful check amount of groups', async () => {
```
Comment (masteralx) [inline test/4_QuestionsWithGroups.spec.js]: successfully

File: test/4_QuestionsWithGroups.spec.js [hunk 24]
```diff
@@ -33,10 +35,87 @@ contract('QuestionsWithGroups', (accounts) => {
...
+
+    describe('addQuestionGroup()', () => {
+        it('should successful add group', async () => {
+            const tx = await questionsWithGroups.addQuestionGroup(group);
+            const uploadedGroup = await questionsWithGroups.getQuestionGroup(1);
+            assert.strictEqual(uploadedGroup.name , 'group name');
+
+            const log = tx.logs.find(element => element.event.match('QuestionGroupAdded'));
+            const {args: {name}} = log;
+            assert.strictEqual(name, group.name);
```
Comment (masteralx) [inline test/4_QuestionsWithGroups.spec.js]: separate test

File: test/6_Controlled.spec.js [hunk 25]
```diff
@@ -32,11 +32,44 @@ contract('ZeroOne', (accounts) => {
     beforeEach(async () => {
         zeroOne = await ZeroOne.new({ from: deployFrom });
         controlled = await Controlled.new(zeroOne.address, { from: deployFrom });
+        //console.log(controlledMethods);
```
Comment (masteralx) [inline test/6_Controlled.spec.js]: remove

**Pull request merged**